File: src/fs/dir.rs
/*
 MIT License

 Copyright (c) 2019 Max Kuznetsov <syhpoon@syhpoon.ca>

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
*/

use std::convert::TryFrom;
use std::collections::HashMap;

use failure::Error;

use crate::device::Device;
use super::{Filesystem, Inode};
use super::types::InodePtr;
use super::dir_entry::DirEntry;
use super::types::BlockPtr;

pub struct Directory {
    name2inode: HashMap<String, InodePtr>,
    entries: Vec<(Option<DirEntry>, BlockPtr)>,
}

impl Directory {
    /// Load directory entries
    pub fn load<T: Device>(inode: &Inode,
                           fs: &mut Filesystem<T>) -> Result<Self, Error> {
        let mut dir = Directory {
            name2inode: HashMap::new(),
            entries: vec![],
        };

        // Load every assigned block and scan for dir entries
        for bid in inode.get_blocks() {
            let bcell = fs.get_block(bid)?;
            let block = bcell.borrow();

            let mut start = 0;
            let mut end = DirEntry::SIZE;

            while end < block.size() {
                let slice = &block.data[start..end];
                let entry = DirEntry::try_from(slice)?;

                // inode == 0 indicates unused entry
                if entry.inode == 0 {
                    dir.entries.push((None, bid));
                } else {
                    dir.name2inode.insert(entry.name_as_string(), entry.inode);
                    dir.entries.push((Some(entry), bid));
                }

                start += DirEntry::SIZE;
                end += DirEntry::SIZE;
            }
        }

        Ok(dir)
    }

    pub fn add_entry<T: Device>(&mut self, name: String,
                                inode: InodePtr,
                                fs: &mut Filesystem<T>) -> Result<(), Error> {
        let entry = DirEntry::new(inode, name.as_str())?;
        self.name2inode.insert(name, inode);

        // Find the first empty entry slot
        match self.entries.iter().position(|e| e.0.is_none()) {
            Some(idx) => {
                let bid = self.entries[idx].1;
                let offset = idx * DirEntry::SIZE;
                let bytes: Vec<u8> = bincode::serialize(&entry)?;
                let range = offset..(offset + DirEntry::SIZE);
                let bcell = fs.get_block(bid)?;
                let mut block = bcell.borrow_mut();

                // Replace serialized entry
                block.data.splice(range, bytes);
                block.set_dirty(true);

                self.entries[idx] = (Some(entry), bid)
            }
            None => unimplemented!("TODO: allocate more blocks")
        };

        Ok(())
    }
}

File: pkg/fs/dir.go
// Copyright (c) 2022 Max Kuznetsov <syhpoon@syhpoon.ca>

package fs

import (
	"bytes"
	"fmt"
	"sync"
	"sync/atomic"

	"github.com/syhpoon/learnfs/pkg/device"
)

type Directory struct {
	inode      *Inode
	name2inode map[string]InodePtr
	entries    map[BlockPtr][]*DirEntry
	cache      *blockCache
	alloc      BlockAllocator
	pool       *bufPool
	blockSize  uint32
	dirty      int32

	sync.RWMutex
}

func NewDirectory(inode *Inode, cache *blockCache, pool *bufPool, blockSize uint32) *Directory {
	entries := make(map[BlockPtr][]*DirEntry)

	// For every inode block create a slice of nil dir entries to represent
	for _, blkPtr := range inode.getBlockPtrs() {
		entries[blkPtr] = make([]*DirEntry, blockSize/DIR_ENTRY_SIZE)
	}

	dir := &Directory{
		inode:      inode,
		name2inode: make(map[string]InodePtr),
		entries:    entries,
		cache:      cache,
		blockSize:  blockSize,
		pool:       pool,
	}

	return dir
}

func LoadDirectory(inode *Inode, cache *blockCache,
	pool *bufPool, blockSize uint32) (*Directory, error) {

	dir := NewDirectory(inode, cache, pool, blockSize)

	// For every block load the corresponding dir entries
	for _, blkPtr := range inode.getBlockPtrs() {
		blk, err := cache.getBlock(blkPtr)
		if err != nil {
			return nil, fmt.Errorf("failed to load block %d: %w", blkPtr, err)
		}

		dir.entries[blkPtr] = make([]*DirEntry, dir.blockSize/DIR_ENTRY_SIZE)

		// Load all the dir entries in the block
		start := 0
		end := DIR_ENTRY_SIZE

		blk.Lock()
		buf := blk.data

		entryIdx := 0
		for end < blk.size() {
			slice := buf[start:end]
			entry := &DirEntry{}

			// TODO: add direntry magic and check it here
			if err := entry.DecodeFrom(bytes.NewReader(slice)); err != nil {
				blk.Unlock()
				return nil, fmt.Errorf("failed to load dir entry: %w", err)
			}

			if entry.InodePtr == 0 {
				dir.entries[blkPtr][entryIdx] = nil
			} else {
				dir.entries[blkPtr][entryIdx] = entry
				dir.name2inode[entry.GetName()] = entry.InodePtr
			}

			start += DIR_ENTRY_SIZE
			end += DIR_ENTRY_SIZE
			entryIdx++
		}

		blk.Unlock()
	}

	return dir, nil
}

func (dir *Directory) AddEntry(name string, inodePtr InodePtr) error {
	dir.Lock()
	defer dir.Unlock()

	if _, ok := dir.name2inode[name]; ok {
		return ErrorAlreadyExists
	}

	entry, err := NewDirEntry(inodePtr, name)
	if err != nil {
		return fmt.Errorf("failed to create new dir entry: %w", err)
	}

	dir.name2inode[name] = inodePtr

	for _, blkPtr := range dir.inode.getBlockPtrs() {
		// Find the first empty slot
		for i := range dir.entries[blkPtr] {
			if dir.entries[blkPtr][i] == nil {
				dir.entries[blkPtr][i] = entry
				dir.SetDirty(true)

				return nil
			}
		}
	}

	return fmt.Errorf("TODO: allocate more directory blocks")
}

func (dir *Directory) flush(dev device.Device, sb *Superblock) error {
	dir.Lock()
	defer dir.Unlock()

	emptyEntry := &DirEntry{}

	write := func(ptr BlockPtr, entries []any) error {
		bin, err := serialize(dir.pool, entries...)
		if err != nil {
			return fmt.Errorf("failed to serialize dir entries: %w", err)
		}

		op := &device.Op{
			Buf:    bin,
			Offset: sb.BlockOffset(ptr),
		}

		if err := dev.Write(op); err != nil {
			return fmt.Errorf("failed to write dir entries: %w", err)
		}

		return nil
	}

	var entries []any
	for _, blkPtr := range dir.inode.getBlockPtrs() {
		for _, e := range dir.entries[blkPtr] {
			entry := e
			if entry == nil {
				entry = emptyEntry
			}

			entries = append(entries, entry)
		}

		if err := write(blkPtr, entries); err != nil {
			return err
		}

		entries = entries[:0]
	}

	return nil
}

func (dir *Directory) GetEntry(name string) (InodePtr, error) {
	dir.RLock()
	ptr, ok := dir.name2inode[name]
	dir.RUnlock()

	if !ok {
		return 0, ErrorNotFound
	}

	return ptr, nil
}

func (dir *Directory) DeleteEntry(name string) error {
	dir.Lock()
	delete(dir.name2inode, name)

	for _, entries := range dir.entries {
		for j, entry := range entries {
			if entry != nil {
				if entry.GetName() == name {
					entries[j] = nil
				}
			}
		}
	}
	dir.SetDirty(true)
	dir.Unlock()

	return nil
}

func (dir *Directory) GetEntries() []*DirEntry {
	dir.RLock()
	defer dir.RUnlock()

	var entries []*DirEntry

	for blkPtr := range dir.entries {
		for i := range dir.entries[blkPtr] {
			if dir.entries[blkPtr][i] == nil {
				continue
			}

			e := *dir.entries[blkPtr][i]
			entries = append(entries, &e)
		}
	}

	return entries
}

func (dir *Directory) SetDirty(dirty bool) {
	val := int32(1)
	if !dirty {
		val = 0
	}

	atomic.StoreInt32(&dir.dirty, val)
}

func (dir *Directory) IsDirty() bool {
	return atomic.LoadInt32(&dir.dirty) == 1
}
